crates/commonware-node/src/dkg/manager/actor/state.rs

```diff
@@ -1,140 +1,1092 @@
+const PAGE_SIZE: NonZeroUsize = NZUsize!(1 << 12);
+const POOL_CAPACITY: NonZeroUsize = NZUsize!(1 << 20);
+const WRITE_BUFFER: NonZeroUsize = NZUsize!(1 << 12);
+const READ_BUFFER: NonZeroUsize = NZUsize!(1 << 20);
 
-/// All state for an epoch:
+/// The maximum number of validators ever permitted in the DKG ceremony.
 ///
-/// + the DKG outcome containing the public key, the private key share, and the
-///   participants for the epoch
-/// + the validator state, containing the dealers of the epoch (corresponds to
-///   the participants in the DKG outcome), the players of the next ceremony,
-///   and the syncing players, who will be players in the ceremony thereafter.
-#[derive(Clone, Debug)]
-pub(in crate::dkg::manager) struct State {
-    pub(super) dkg_outcome: DkgOutcome,
-    pub(super) validator_state: ValidatorState,
+/// u16::MAX is 2^16-1 validators, i.e. 65536, which is probably more than
+/// we will ever need. An alternative would be u8::MAX but that feels a bit
+/// too limited. There is extremely little cost doing u16::MAX instead.
+const MAXIMUM_VALIDATORS: NonZeroU32 = NZU32!(u16::MAX as u32);
+
+pub(super) fn builder() -> Builder {
+    Builder::default()
 }
 
-impl State {
-    pub(crate) fn epoch(&self) -> Epoch {
-        self.dkg_outcome.epoch
+pub(super) struct Storage<TContext>
+where
+    TContext: commonware_runtime::Storage + Metrics,
+{
+    states: contiguous::variable::Journal<TContext, State>,
+    events: segmented::variable::Journal<TContext, Event>,
+
+    current: State,
+    cache: BTreeMap<Epoch, Events>,
+}
+
+impl<TContext> Storage<TContext>
+where
+    TContext: commonware_runtime::Storage + Metrics,
+{
+    /// Returns all player acknowledgments received during the given epoch.
+    fn acks_for_epoch(
+        &self,
+        epoch: Epoch,
+    ) -> impl Iterator<Item = (&PublicKey, &PlayerAck<PublicKey>)> {
+        self.cache
+            .get(&epoch)
+            .into_iter()
+            .flat_map(|cache| cache.acks.iter())
+    }
+
+    /// Returns all dealings received during the given epoch.
+    fn dealings_for_epoch(
+        &self,
+        epoch: Epoch,
+    ) -> impl Iterator<Item = (&PublicKey, &(DealerPubMsg<MinSig>, DealerPrivMsg))> {
+        self.cache
+            .get(&epoch)
+            .into_iter()
+            .flat_map(|cache| cache.dealings.iter())
+    }
+
+    /// Returns all dealings received during the given epoch.
+    pub(super) fn logs_for_epoch(
+        &self,
+        epoch: Epoch,
+    ) -> impl Iterator<Item = (&PublicKey, &dkg::DealerLog<MinSig, PublicKey>)> {
+        self.cache
+            .get(&epoch)
+            .into_iter()
+            .flat_map(|cache| cache.logs.iter())
+    }
+
+    /// Returns the DKG outcome for the current epoch.
+    pub(super) fn current(&self) -> State {
+        self.current.clone()
+    }
+
+    /// Appends the outcome of a DKG ceremony to state
+    pub(super) async fn append_state(&mut self, state: State) -> eyre::Result<()> {
```

Comment:
should we clear out the cache from previous epoch(s) here? i didnt see any clean-up i think